File: cmps/101/programming/pa2/apint.c
#include <stdio.h>
#include <stdlib.h>
#include <inttypes.h>
#include <string.h>
#include <ctype.h>
#include <assert.h>
#include "apint.h"

#define FALSE 0
#define TRUE 1

/* Private definitions */

typedef struct apintObj {
    uint8_t *data;
    size_t size;
    uint8_t isNeg;
} apintObj;

/*
  Compares a to b
  Input: apint a, apint b
  Output: -1 if a < b, 0 if a == b, 1 if a > b
 */
int8_t compareTo(apint a, apint b) {
    if (a->isNeg != b->isNeg)
        return a->isNeg ? -1 : 1;
    if (a->size != b->size) {
        if (a->isNeg)
            return b->size > a->size ? 1 : -1;
        return a->size > b->size ? 1 : -1;
    }
    for (size_t i = a->size - 1; i != -1; i--)
        if (a->data[i] != b->data[i])
            return a->data[i] > b->data[i] ? 1 : -1;
    return 0;
}

/*
  Clones an existing apint
  Input: apint
  Output: apint
 */
apint newApintApint(apint a, size_t dataSize) {
    apint ret = malloc(sizeof *ret);
    assert(ret != NULL);

    ret->size = a->size;
    ret->isNeg = a->isNeg;

    ret->data = calloc(dataSize, sizeof *(ret->data));
    assert(ret->data != NULL);
    memcpy(ret->data, a->data, ret->size);

    return ret;
}

/* Public definitions */

/*
  Returns an apint set to 0 with the given size
  Input: size_t
  Output: apint
 */
apint newApint(size_t size) {
    apint ret = malloc(sizeof *ret);
    assert(ret != NULL);

    ret->size = 1;
    ret->isNeg = FALSE;

    ret->data = calloc(size, sizeof *(ret->data));
    assert(ret->data != NULL);

    return ret;
}

/*
  Creates a new apint from a strin gofmatted according to regex "[+-]?[0-9]+"
  Input: string
  Output: apint
 */
apint newApintStr(char *val) {
    size_t strsize = strlen(val);

    apint ret = malloc(sizeof *ret);
    assert(ret != NULL);

    ret->size = strsize - !isdigit(val[0]);

    ret->data = calloc(strsize, sizeof *(ret->data));
    assert(ret->data != NULL);

    ret->isNeg = (!isdigit(val[0]) && (val[0] == '-'));

    for (size_t i = strsize - 1; i != -1 && isdigit(val[i]); i--)
        ret->data[strsize - i - 1] = val[i] - '0';
    return ret;
}

/*
  Creates a new apint from an integer
  Input: int
  Output: apint
 */
apint newApintInt(int val) {
    int i;
    apint ret = calloc(1, sizeof *ret);
    assert(ret != NULL);

    if (val == 0) {
        ret->data = calloc(1, sizeof *(ret->data));
        assert(ret->data != NULL);
        ret->size = 1;
        ret->isNeg = FALSE;
        return ret;
    } else {
        ret->isNeg = (val < 0);
        if (ret->isNeg)
            val *= -1;
        for (i = val; i != 0; i /= 10, ret->size++);

        ret->data = calloc(ret->size, sizeof *ret->data);
        assert(ret->data != NULL);
        for(i = 0; i < ret->size; i++) {
            ret->data[i] = val % 10;
            val /= 10;
        }
    }

    return ret;
}

/*
  Frees all data allocated to an apint. Also sets that apint poter to NULL
  Input: apint
  Output: void
 */
void freeApint(apint *a) {
    if (a != NULL && *a != NULL) {
        if ((*a)->data != NULL)
            free((*a)->data);
        free(*a);
        *a = NULL;
    }
}

/*
  Adds 2 apints together and returns the result.
  Input: apint, apint
  Output: apint
 */
apint add(apint a, apint b) {
    apint ret, other;
    uint8_t aSign = a->isNeg;
    uint8_t bSign = b->isNeg;
    int8_t carry = 0;
    size_t i;

    // Alias a and b in order to not have to check sign and length. ret= bigger number (absolute size).
    a->isNeg = FALSE;
    b->isNeg = FALSE;
    if (compareTo(a, b) >= 0) {
        ret = newApintApint(a, a->size + 1);
        other = b;
        ret->isNeg = aSign;
        b->isNeg = bSign;
    } else {
        ret = newApintApint(b, b->size + 1);
        other = a;
        ret->isNeg = bSign;
        other->isNeg = aSign;
    }
    a->isNeg = aSign;
    b->isNeg = bSign;

    // If both have the same sign, add with carry
    if (ret->isNeg == other->isNeg) {
        for (i = 0; i < ret->size; i++) {
            if (i >= other->size)
                carry += ret->data[i];
            else
                carry += ret->data[i] + other->data[i];
            ret->data[i] = carry % 10;
            carry /= 10;
        }
        if (carry != 0) {
            ret->data[i+1] = carry;
            ret->size++;
        }
    } else {
        // If both have different signs, subtract with borrowing
        for (i = 0; i < ret->size; i++) {
            if (i >= other->size)
                carry += ret->data[i];
            else
                carry += ret->data[i] - other->data[i];
            if (carry < 0) {
                ret->data[i] = carry + 10;
                carry = -1;
            } else {
                ret->data[i] = carry;
                carry = 0;
            }
        }
        while (ret->size > 0 && ret->data[ret->size-- - 1] == 0);
        ret->size++;
    }

    return ret;
}

/* Subtracts b from a  */
apint subtract(apint a, apint b) {
    b->isNeg = !b->isNeg;
    apint ret = add(a, b);
    b->isNeg = !b->isNeg;
    return ret;
}

/*
  Multiply two apints together
  Input: apint
  Output apint
*/
apint multiply(apint a, apint b) {
    apint ret, temp, me, other, holder;
    uint8_t carry, currentDigit;
    size_t i, j, maxSize;

    // Set me to the apint that has the most digits
    if (a->size > b->size) {
        me = a;
        other = b;
    } else {
        me = b;
        other = a;
    }

    // Initialize variables
    maxSize = me->size + other->size + 1;
    ret = newApint(maxSize);
    temp = newApint(maxSize);

    for (i = 0; i < other->size; i++) {
        carry = 0;
        memset(temp->data, 0, maxSize);
        currentDigit = other->data[i];

        // Multiply
        for (j = 0; j < me->size; j++) {
            carry += currentDigit * me->data[j];
            temp->data[i+j] = carry % 10;
            carry /= 10;
        }
        temp->size = i+j+1;
        if (carry != 0)
            temp->data[i+j] = carry;

        // Add intermidiate results
        holder = ret;
        ret = add(ret, temp);
        freeApint(&holder);
    }

    // Set the sign of the new number. Ensure that 0 is always positive
    if (ret->size == 1 && ret->data[0] == 0)
        ret->isNeg = FALSE;
    else
        ret->isNeg = other->isNeg ^ me->isNeg;

    while (ret->size > 0 && ret->data[ret->size-- - 1] == 0);
    ret->size++;

    return ret;
}

void print(apint a) {
    if (a->isNeg)
        putchar('-');
    for (size_t i = a->size - 1; i != -1; i--)
        putchar(a->data[i] + '0');
}

File: cmps/101/programming/pa1/apint.java
/*
  CREATED: Ilay Raz 4/20/18
  apint class file
 */
import java.util.*;

class apint {
    // Declare some constants
    public static final apint ZERO = new apint(0);
    public static final apint ONE = new apint(1);

    private ArrayList<Byte> data; // Each byte stores one digit
    private boolean isNeg; // Stores sign

    /*
       Creates a new apint initalized to value 0
       Input: null
       Output: apint object
    */
    public apint() {
        this.data = new ArrayList<Byte>();
        this.addDigit((byte)0);
        this.isNeg = false;
    }

    /*
      Creates a new apint from an integer
      Input: int
      Output: apint object
    */
    public apint(int num) {
        this.data = new ArrayList<Byte>();
        this.isNeg = Integer.signum(num) == -1;
        if (num == 0)
            this.addDigit((byte)0);
        else {
            num = this.isNeg ? num*-1 : num;
            while (num != 0) {
                this.addDigit((byte)(num % 10));
                num /= 10;
            }
        }
    }

    /*
      Creates a new apint from a double
      Input: double
      Output: apint object
    */
    public apint(double num) {
        this(String.format("%.0f", num));
    }

    /*
      Creates a new apint from a string formatted according to regex "[+-]?[0-9]+"
      Input: string formatted according to regex "[+-]?[0-9]+"
      Output: apint object
    */
    public apint(String num) {
        this.data = new ArrayList<Byte>(num.length());

        int endPoint = Character.isDigit(num.charAt(0)) ? 0 : 1;
        for (int i = num.length() - 1; i >= endPoint; i--)
            this.addDigit((byte)Character.digit(num.charAt(i), 10));

        if (this.digitCount() == 1 && this.getDigit(0) == (byte)0)
            this.isNeg = false;
        else
            this.isNeg = num.charAt(0) == '-' ? true : false;
    }

    /*
      Clones an existing apint
      Input: apint object
      Output: apint object
    */
    public apint(apint num) {
        this.data = new ArrayList<Byte>(num.data);
        this.isNeg = num.isNeg;
    }

    /*
      Adds 2 apints together and returns the result. Does not modify the original apint.
      Input: apint to be added
      Output: apint resulted
     */
    public apint add(apint num) {

        // Alias this and num in order to not have to check sign and length. Set ret=bigger number
        apint ret, other;
        boolean thisSign = this.getSign();
        boolean numSign = num.getSign();
        this.setSign(false);
        num.setSign(false);
        if (this.compareTo(num) >= 0) {
            ret = new apint(this);
            other = num;
            ret.setSign(thisSign);
            other.setSign(numSign);
        } else {
            ret = new apint(num);
            other = this;
            ret.setSign(numSign);
            other.setSign(thisSign);
        }
        this.setSign(thisSign);
        num.setSign(numSign);

        byte carry = 0;

        // If both have the same sign, add with carry
        if (ret.isNeg == other.isNeg) {
            for (int i = 0; i < ret.digitCount(); i++) {
                if (i >= other.digitCount())
                    carry += ret.getDigit(i);
                else
                    carry += ret.getDigit(i) + other.getDigit(i);
                ret.setDigit(i, (byte)(carry % 10));
                carry /= 10;
            }
            if (carry != 0)
                ret.addDigit(carry);
        } else {
            // If both have different signs, subtract with borrowing (don't have to worry about signs because ret is always the bigger number
            for (int i = 0; i < ret.digitCount(); i++) {
                if (i >= other.digitCount())
                    carry += ret.getDigit(i);
                else
                    carry += ret.getDigit(i) - other.getDigit(i);
                if (carry < 0) {
                    ret.setDigit(i,(byte)(carry + 10));
                    carry = -1;
                } else {
                    ret.setDigit(i, carry);
                    carry = 0;
                }
            }
        }

        return ret;
    }

    /*
      Multiply two apints together
      Input: apint
      Output: apint whose value is (this*num)
     */
    public apint multiply(apint num) {
        apint ret, temp, me, other;
        byte carry, currentDigit;

        // Set me to the apint that has the most digits
        if (this.digitCount() > num.digitCount()) {
            me = this;
            other = num;
        } else {
            me = num;
            other = this;
        }

        // Initialize variables
        ret = new apint();
        temp = new apint();
        ret.data.ensureCapacity(me.digitCount() + other.digitCount());
        temp.data.ensureCapacity(me.digitCount() + other.digitCount());

        for (int i = 0; i < other.digitCount(); i++) {
            carry = 0;
            temp.data.clear();
            currentDigit = other.getDigit(i);

            // Ensure significant digit placment is corrent
            for (int j = 0; j < i; j++)
                temp.addDigit((byte)0);

            // Multiply
            for (int j = 0; j < me.digitCount(); j++) {
                carry += currentDigit * me.getDigit(j);
                temp.addDigit((byte)(carry % 10));
                carry /= 10;
            }
            if (carry != 0)
                temp.addDigit((byte)carry);

            // Add intermidiate result
            ret = ret.add(temp);
        }

        // Set the sign of the new number. Ensure that 0 is always positive
        if (ret.digitCount() == 1 && ret.getDigit(0) == (byte)0)
            ret.isNeg = false;
        else
            ret.isNeg = other.isNeg ^ me.isNeg;
        return ret;
    }

    /*
      Subtracts two apints by calling add on the negative of the argument.
      Input: apint
      Output: apint
     */
    public apint subtract(apint num) {
        apint temp = new apint(num);
        temp.isNeg = !temp.isNeg;
        return this.add(temp);
    }

    /*
      Divides this by num using the subtraction method and returns an apint array [div, remainder]
      Input: apint
      Output: apint[] = {div, remainder}
    */
    public apint[] divmod(apint num) throws ArithmeticException {
        if (num.compareTo(apint.ZERO) == 0)
            throw new ArithmeticException("Division by 0");

        apint remainder = new apint(this);
        apint numPositive = new apint(num);
        remainder.isNeg = numPositive.isNeg = false;
        apint quotient = new apint();
        ArrayDeque<Byte> Dquotient = new ArrayDeque<Byte>();
        apint temp = new apint();
        byte counter = 0;
        temp.data.ensureCapacity(num.digitCount());

        for (int i = this.digitCount() - num.digitCount(); i >= 0; i--) {
            temp.data.clear();
            counter = 0;

            // Align numbers
            for (int j = 0; j < i; j++)
                temp.addDigit((byte)0);
            temp.data.addAll(numPositive.data);

            // Subtract
            while (remainder.compareTo(temp) >= 0) {
                remainder = remainder.subtract(temp);
                counter++;
            }
            Dquotient.addFirst(counter);
        }

        /*while (remainder.compareTo(numPositive) >= 0) {
            ret = ret.add(apint.ONE);
            remainder = remainder.subtract(numPositive);
            }*/

        quotient.data = new ArrayList<Byte>(Dquotient);

        if (quotient.compareTo(apint.ZERO) != 0)
            quotient.isNeg = this.isNeg ^ num.isNeg;

        apint[] retArray = {quotient, remainder};
        return retArray;
    }

    /*
      Divides this by num. Throws an error if modulus is not 0
      Input: apint
      Output: apint
     */
    public apint divide(apint num) throws ArithmeticException {
        apint[] temp = this.divmod(num);

        if (temp[1].compareTo(apint.ZERO) != 0)
            throw new ArithmeticException("Numbers are not divisible");
        return temp[0];
    }

    /*
      Find modulus of this and num.
      Input: apint
      Output: apint
     */
    public apint modulus(apint num) {
        return this.divmod(num)[1];
    }

    /*
      Compares this to num.
      Input: apint
      Output: -1 if this < num, 0 if this == num, 1 if this > num
     */
    public int compareTo(apint num) {
        if (this.isNeg != num.isNeg)
            return this.isNeg ? -1 : 1;
        if (this.digitCount() != num.digitCount()) {
            if (this.isNeg)
                return num.digitCount() > this.digitCount() ? 1 : -1;
            return this.digitCount() > num.digitCount() ? 1 : -1;
        }
        for (int i = this.digitCount()-1; i >= 0; i--)
            if (this.getDigit(i) != num.getDigit(i))
                return this.getDigit(i) > num.getDigit(i) ? 1 : -1;
        return 0;
    }

    /*
      Returns string representation of apint.
      Input: none
      Output: String
     */
    public String toString() {
        StringBuilder builder = new StringBuilder(this.digitCount() + 1);
        for (int i = 0; i < this.digitCount(); i++)
            builder.append(this.getDigit(i));
        builder.append(this.isNeg ? "-" : "");
        return builder.reverse().toString();
    }

    /*
      Returns the absolute value of the apint.
      Input: none
      Output: apint
     */
    public apint abs() {
        apint temp = new apint(this);
        temp.isNeg = false;
        return temp;
    }

    /*
      Returns the sign of the apint. true = negative, false = positive.
      Input: none
      Output: boolean
    */
    public boolean getSign() {
        return this.isNeg;
    }

    public void setSign(boolean sign) {
        this.isNeg = sign;
    }

    public void print() {
        System.out.println(this.toString());
    }

    /*
      Returns number of digits
      Input: none
      Output: int
     */
    private int digitCount() {
        int max;
        for (max = this.data.size(); max > 1 && this.getDigit(max-1) == (byte)0; max--);
        return max;
    }

    /*
      Get a digit at specified index
      Input: int
      Output: byte
     */
    private byte getDigit(int i) {
        return this.data.get(i);
    }

    /*
      Appends a digit
      Input: byte
      Output: none
     */
    private void addDigit(byte val) {
        this.data.add(val);
    }

    /*
      Sets a digit at index to value
      Input: int index, byte value
      Output: none
     */
    private void setDigit(int i, byte val) {
        this.data.set(i, val);
    }
}
